### ISSUE TITLE
[rootcling] `genreflex --cxxmodule` generates a truncated .pcm file

### ISSUE THREAD

**Issue by jalopezg-git** [root-project/root#9448]
- [X] Checked for duplicates

### Describe the bug
`genreflex --cxxmodule ...` generates a truncated `xxx_rflx.pcm` file (0 bytes).  The equivalent invocation of `rootcling --cxxmodule` using a linkdef.h instead of a XML file works as expected.

### Expected behavior
`genreflex` should generate a valid cxxmodule file.

### To Reproduce
Run genreflex/rootcling as follows.  The contents of the required files are provided below.
```bash
$ genreflex --cxxmodule -s classes.xml classes.h
$ # or
$ rootcling --cxxmodule -f out.cpp -s ./libclasses.so classes.h classes_linkdef.h
```

classes.h:
```c++
#ifndef CLASSES_H
#define CLASSES_H

struct Foo {
   Foo(int _i) : i(_i) {}
   int i = 0;
};

#endif
```

classes_linkdef.h:
```c++
#ifdef __CINT__
#pragma link off all globals;
#pragma link off all classes;
#pragma link off all functions;

#pragma link C++ class Foo+;
#endif
```

classes.xml:
```xml
<lcgdict>
  <class name="Foo"/>
</lcgdict>
```

### Setup
ROOT master, Arch Linux x86_64

**Comment by Axel-Naumann** [comment]
IIRC that's expected (@vgvassilev might correct me): genreflex embeds the header as string in the dictionary. As such we don't need an actual PCM file; we provide one to regularize the build systems.

**Comment by pcanal** [comment]
humm ... why?  Is that much much slower (parsing) than a binary pcm or do you mean that it is a binary blob?  Either way why is the option --cxxmodule behaving differently for genreflex than rootcling?

**Comment by vgvassilev** [comment]
On my machine both commands produce a 0-sized module. Eg:
```bash
rootcling --cxxmodule -f out.cpp -s ./libclasses.so classes.h classes_linkdef.h
Error in <CheckModuleValid>: Couldn't find module with name 'classes' in modulemap!
du classes.pcm 
0	classes.pcm
```

That is because you need to write a modulemap (normally the build system or aclic does it for you).

```bash
echo "module classes { export * header \"classes.h\" }" > classes.modulemap
rootcling -cxxmodule -moduleMapFile=classes.modulemap  -f out.cpp -s ./libclasses.so  classes.h classes_linkdef.h 
vvassilev@vv-nuc ~/workspace/builds/root_master $ du -hs classes.pcm 
400K	classes.pcm
```

If you want to use the selection xml file you could still do it like:
```bash
rootcling -reflex -cxxmodule -moduleMapFile=classes.modulemap  -f out.cpp -s ./libclasses.so  classes.h classes.xml 
du classes.pcm 
404	classes.pcm
```

We never really tried the `-cxxmodule` option under `genreflex`.  Maybe we should just drop it from there. We use `rootcling -reflex ...`. See here https://github.com/cms-sw/cmssw-config/blob/41730a3bd862bb50a68d52279d096f1794d8874e/SCRAM/GMake/Makefile.cxxmodule#L112
Using `rootcling -reflex` would not follow the `genreflex` convention to add `_rflx` prefix to the files but so far we have not seen issues due to that.

A general note is that `rootcling` is a superset of `genreflex`. The argument parsing of rootcling is more helpful and we run extra checks in the context of `-cxxmodule`. I'd warmly recommend to use rootcling in favor of genreflex in the context of modules. Additionally, this gives a potential way to deprecate `genreflex` once/if people move to modules.

**Comment by dpiparo** [comment]
This can be fixed working on the commandline arguments. Presenly the `-moduleMapFile...` flag passed to genreflex is translated to `-m oduleMapFile...` to rootcling, which is the bug in the current master/6.30-patches. Assigning to me.

**Comment by vgvassilev** [comment]
We should probably deprecate genreflex in favor of rootcling, as there will always be such fallout. We should probably advertise `genreflex --debug` to get the right rootcling arguments for those who want to migrate and use modules.

**Comment by dpiparo** [comment]
I support this kind of simplifications: deprecating genreflex is an option. Before discussing that in detail, the advantages of the rootcling only approaches should be listed carefully. Users such as CMS and LHCb will have to migrate away of genreflex.

**Comment by vgvassilev** [comment]
I believe CMS has done it in the CXXMODULES IB and IIRC @smuzaffar planned to put it in the mainline...

**Comment by smuzaffar** [comment]
Yes @vgvassilev , for CMS we are already using `rootcling`. The `genreflex` command we normally run is
```
genreflex classes.h -s classes_def.xml 
-o LibraryName_xr.cc 
--fail_on_warnings 
--rootmap=LibraryName_xr.rootmap 
--rootmap-lib=libLibraryName.so 
-m DepsLibrary1_xr_rdict.pcm -m DepsLibrary2_xr_rdict.pcm 
<CPPFLAGS>
-I<include-dirs>
```

and for CXXMODULE builds we run 
```
rootcling -reflex -f LibraryName_xr.cc -s libLibraryName.so -cxxmodule -failOnWarnings 
-m DepsLibrary1_xr_rdict.pcm -m DepsLibrary2_xr_rdict.pcm
-moduleMapFile=<path/module.modulemap> -moduleMapFile=<path/module.modulemap 
<CPPFLAGS>
-I<include-dirs>
classes.h classes_def.xml
```

I guess, for non-cxxmocule build I can just drop `-cxxmodule` and `-moduleMapFile`. What about `--rootmap=LibraryName_xr.rootmap --rootmap-lib=libLibraryName.so` which we pass to `genreflex`?

**Comment by vgvassilev** [comment]
> What about `--rootmap=LibraryName_xr.rootmap --rootmap-lib=libLibraryName.so` which we pass to `genreflex`?

I can think of it from the top of my head but if you append `--debug` to `genreflex` it would give you what you need to call `rootcling` I think...

**Comment by smuzaffar** [comment]
@vgvassilev , source generated directly with `rootcling` failed for cmssw see https://github.com/cms-sw/cmsdist/issues/8983  for details

**Comment by krasznaa** [comment]
> I support this kind of simplifications: deprecating genreflex is an option. Before discussing that in detail, the advantages of the rootcling only approaches should be listed carefully. Users such as CMS and LHCb will have to migrate away of genreflex.

Que?

We have thousands and thousands of `selection.xml` files in ATLAS, many of which contain very non-trivial definitions. 🤔 Hopefully nobody is seriously suggesting that this language of describing the work for the dictionary generator is in danger. As in ATLAS this would require a full work-plan for how we **might** be able to do it before LHC's Run-4.

**Comment by pcanal** [comment]
> We have thousands and thousands of selection.xml

rootcling can use selection.xml as is.  the main difference are the command line argument spelling (and the options actually available).

**Comment by ferdymercury** [comment]
> the command line argument spelling

See migration helper here https://github.com/root-project/root/pull/16122/files#diff-0a3ef176e860a7266556df1353f14f93dbbb57d500231cd323296f3b8ef2efa7

**Comment by dpiparo** [comment]
Can be closed since rootcling can read xmls since a while.
